use Mod:revert_ordering_on_local_key in {get,delete}_data for DESC keys

hmmr opened the pull request:
Depends on https://github.com/basho/riak_ql/pull/162.

Single-key primitive operations in `riak_kv_ts_api` need to provide to `riak_client:get` and `:delete` keys adjusted for fields declared DESC in the DDL, as is done for list_keys in https://github.com/basho/riak_kv/blob/riak_ts-develop/src/riak_kv_qry_worker.erl#L190.

This fixes regressions currently seen in `ts_simple_single_key_ops` (and possibly others).

Commit:
use riak_ql_ddl:desc_adjusted_key in {get,delete}_data for DESC keys

davidx commented:
thumbot retry

davidx commented:
thumbot retry

Commit:
less list-tuple conversions in calls to riak_ql_ddl:lk_to_pk, :desc_adjusted_key

Commit:
be DESC-aware in qry_worker:make_key_conversion_fun

Commit:
rename desc_adjusted_key to negate_if_desc, rearrange params in calls to lk_to_pk

Commit:
whitespace cleanup

Commit:
use Mod:revert_ordering_on_local_key/1 instead of riak_ql_ddl:negate_if_desc

Commit:
Merge branch 'riak_ts-develop' into fix-az-single_key_ops_desc_keys_regression

macintux commented on src/riak_kv_qry_worker.erl:
```diff
@@ -187,10 +187,8 @@ make_key_conversion_fun(Table) ->
-                riak_ql_ddl:negate_if_desc(
-                  tuple_to_list(sext:decode(Key)), Mod, DDL),
-            {ok, PK} = riak_ql_ddl:lk_to_pk(DescAdjustedKey, Mod, DDL),
+            {ok, PK} = riak_ql_ddl:lk_to_pk(
```
Where will the `bad_compound_key` exception be handled in this case?

hmmr replied:
Over there: https://github.com/basho/riak_kv/pull/1560/commits/b6f142e66433d844fc01d3e982546fa2bca1c63f#diff-47d6b445c0ee52daa420d1d48d98f394R638

macintux replied:
This function is passed from a `riak_kv_index_fsm` process to a `riak_kv_vnode` to be invoked in a fold. I don't see any exception handling in a quick scan of that code path, which leads me to believe that whatever provokes that exception could bring down a vnode.

hmmr replied:
In contrast to its use in `riak_kv_ts_api:{get,delete}_data` where the key comes from user input, the folding functions here will have the key that's guaranteed to be valid (it is read directly from the backend).

macintux replied:
Ok. Please add a comment here, though, so someone in the future doesn't have the same panic attack.

Commit:
add comments on why lk_to_pk in key conversion funs is safe

Commit:
export & use riak_kv_ts_util:local_to_partition_key as KeyConvFun

Previously, a KeyConvFun used to be constructed on coordinator and sent
to vnodes, where it was passed to backend and used in key filter.  This
works fine as long as vnode and coordinator nodes are of the same version.

This patch, instead, passes a {M, F} tuple, in order for the conversion
function to be constructed on-site (i.e., on the node where it is to be
executed), thus avoiding the dreaded badfun error.

This should allow list_keys to work in mixed-cluster scenarios going
forward from version 1.5.  Note that list_keys will still fail for 1.4+1.5
clusters.

hazen commented:
:+1:

andytill commented:
@thumbot retry

andytill commented:
+1
